jobbergate-api/alembic/versions/20240130_105238--ec2d2948fb41_remap_job_submissions_statuses.py

```diff
@@ -0,0 +1,201 @@
+# revision identifiers, used by Alembic.
+revision = "ec2d2948fb41"
+down_revision = "ba62c0fb9879"
+branch_labels = None
+depends_on = None
+
+
+def upgrade():
+    op.add_column(
+        "job_submissions",
+        sa.Column(
+            "slurm_job_state",
+            sa.Enum(
+                "BOOT_FAIL",
+                "CANCELLED",
+                "COMPLETED",
+                "CONFIGURING",
+                "COMPLETING",
+                "DEADLINE",
+                "FAILED",
+                "NODE_FAIL",
+                "OUT_OF_MEMORY",
+                "PENDING",
+                "PREEMPTED",
+                "RUNNING",
+                "RESV_DEL_HOLD",
+                "REQUEUE_FED",
+                "REQUEUE_HOLD",
+                "REQUEUED",
+                "RESIZING",
+                "REVOKED",
+                "SIGNALING",
+                "SPECIAL_EXIT",
+                "STAGE_OUT",
+                "STOPPED",
+                "SUSPENDED",
+                "TIMEOUT",
+                name="slurmjobstate",
+                native_enum=False,
+            ),
+            nullable=True,
+        ),
+    )
+    op.add_column("job_submissions", sa.Column("slurm_job_info", sa.String(), nullable=True))
+    op.alter_column(
+        "job_submissions",
+        "status",
+        existing_type=postgresql.ENUM(
+            "CREATED",
+            "SUBMITTED",
+            "COMPLETED",
+            "FAILED",
+            "UNKNOWN",
+            "REJECTED",
+            "CANCELLED",
+            name="jobsubmissionstatus",
+        ),
+        type_=sa.Enum(
+            "CREATED",
+            "SUBMITTED",
+            "REJECTED",
+            "DONE",
+            "ABORTED",
+
+            # -- Temporarily included until after remapping --
+            "COMPLETED",
+            "FAILED",
+            "UNKNOWN",
+            "CANCELLED",
+            # ----
+
+            name="jobsubmissionstatus",
+            native_enum=False,
+        ),
+        existing_nullable=False,
+    )
+
+    # -- do the remapping --
+    op.execute("UPDATE job_submissions SET status = 'DONE' WHERE status = 'COMPLETED'")
+    op.execute("UPDATE job_submissions SET status = 'ABORTED' WHERE status = 'FAILED'")
+    op.execute("UPDATE job_submissions SET status = 'ABORTED' WHERE status = 'UNKNOWN'")
+    op.execute("""
+        UPDATE job_submissions
+        SET status = 'ABORTED',
+            slurm_job_state = 'CANCELLED'
+        WHERE status = 'CANCELLED'
+    """)
+    # ----
+
+    op.alter_column(
+        "job_submissions",
+        "status",
+        existing_type=postgresql.ENUM(
+            "CREATED",
+            "SUBMITTED",
+            "REJECTED",
+            "DONE",
+            "ABORTED",
+            "COMPLETED",
+            "FAILED",
+            "UNKNOWN",
+            "CANCELLED",
+            name="jobsubmissionstatus",
+        ),
+        type_=sa.Enum(
+            "CREATED",
+            "SUBMITTED",
+            "REJECTED",
```

Comment:
Do we need three of them? From the agent's perspective, they are all kind of done, it does not need to worry about them anymore. The new columns with slurm status will hold the details differentiating them.

Reply:
I think it's useful to have a distinction in the status for which jobs completed with a normal exit code and those that were not completed for some reason. Also, REJECTED means that the job was never processed by Slurm at all but was rejected at submission time. The statuses have distinct meanings.